### DISCUSSION TITLE
Custom popups

### DISCUSSION THREAD

**Discussion by AshfordN** [fltk-rs/fltk-rs#942] [Q&A] [answered]
Is there a way to implement a custom popup window that perfectly mimics the behavior of a tooltip/menu popup? So far, I have looked at [Erco's example](http://seriss.com/people/erco/fltk/#PopupWindow) as well as the custom_popup example found in this repository, and both examples use separate, frameless windows for the popups; which, incidentally, was also my initial approach. However, there are some minor differences between the results you obtain through this approach and the popups used by tooltips and menus, namely:

1. Clicking outside the popup does not close the popup, but this could be achieved by handling `Unfocus` events
2. Showing the popup window pushes the main window to the background, which does not seem to be the case with tooltip/menu popups; and there is no clear way to remedy this
3. Despite the fact that the popup window is opened and the main window is pushed to the back, you still get visual feedback when hovering over the main window's controls, which does not happen with tooltip/menu popups
4. Closing the main window leaves the popup behind, and the program opened, but this can be solved by setting a callback on the main window that quits the application instead of hiding the main window. Additionally, you can mitigate the issue altogether by setting the window as a modal, but then the workaround in (1) no longer works
5. Frameless windows do not have drop shadows like those those found on tooltip popus, and while it is not impossible to implement, there does not seem to be an easy way of doing it

Below, is the sample code I wrote and an animation demonstrating the differences outlined above.

### main.rs
```rust
use fltk::{app, enums, frame, window};
use fltk::{menu, prelude::*};

fn main() -> Result<(), FltkError> {
    let mut win = window::Window::default()
        .with_size(400, 300)
        .center_screen()
        .with_label("Popup Example");

    let mut activation_frame = frame::Frame::default()
        .with_size(150, 50)
        .center_of_parent()
        .with_label("Left click for popup\nRight click for menu");
    activation_frame.set_tooltip("this is a tooltip");
    activation_frame.set_frame(enums::FrameType::BorderBox);
    win.end();

    let mut popup = window::MenuWindow::default().with_size(150, 60);
    let content = frame::Frame::default()
        .size_of_parent()
        .center_of_parent()
        .with_label("This is a popup");
    popup.end();
    popup.set_border(false);
    popup.set_frame(enums::FrameType::ShadowBox);

    popup.handle(|p, evt| match evt {
        // close the popup when it no longer has focus
        enums::Event::Unfocus => {
            p.hide();
            true
        }
        _ => false,
    });

    activation_frame.handle(move |_, evt| match evt {
        enums::Event::Push => {
            if app::event_mouse_button() == app::MouseButton::Left {
                let (ex, ey) = (app::event_x_root(), app::event_y_root());
                popup.set_pos(ex, ey);
                popup.show();
            } else if app::event_mouse_button() == app::MouseButton::Right {
                let (ex, ey) = app::event_coords();
                let menu = menu::MenuItem::new(&["Item 1", "Item 2"]);
                menu.popup(ex, ey);
            }

            true
        }
        _ => false,
    });

    win.set_callback(|w| {
        if app::event() != enums::Event::Shortcut && app::event_key() != enums::Key::Escape {
            // quit app instead of hiding window
            app::quit();
        }
    });

    win.show();
    app::run()?;
    Ok(())
}
```

### Differences in popups
![demonstration](https://user-images.githubusercontent.com/4706674/137607682-20edb1ee-b465-406b-a000-6392ac3d3f73.gif)
pay attention to the differences in shadow effects and in the behaviour of the main window when the popups are shown.

**Comment by MoAlyousef** [comment]
Theres a 3rd way for popups, by using a MenuButton. 

Regarding mimicking behavior, FLTK passes platform specific flags (like WS_POPUP, WS_EX_TOOLWINDOW on Windows, which can be set using SetWindodowLongA), not sure about what it passes on X11, to get platform like behavior. Windowing (even menus and tooltips) is mostly platform specific, you can still specify a FrameType for a Window and FLTK will try to draw it.  

You can pass these flags manually like here:
https://github.com/fltk-rs/fltk-rs/discussions/837

Or for X11 and Cocoa, the fltk-webview repo does that. 

However to get more info about implementation stuff, I’d suggest you ask on the fltk-general forum. 

https://groups.google.com/g/fltkgeneral

**Comment by AshfordN** [comment, marked as answer]
So I revisited this issue and, it turns out that there is a straightforward way to achieve this using only the FLTK API, as shown in the code below:
```rust
use fltk::{prelude::*, *};

fn main() {
    let root = app::App::default();

    let mut popup = window::MenuWindow::default().with_size(150, 60);
    let mut content = frame::Frame::default()
        .size_of_parent()
        .center_of_parent()
        .with_label("This is a popup");
    content.set_frame(enums::FrameType::BorderBox);
    popup.end();

    // this function sets the NOBORDER and OVERRIDE flags
    // NOTE: this is not currently exposed in the rust binding
    popup.set_override();

    popup.handle(|p, evt| match evt {
        enums::Event::Push => {
            p.hide();
            // stop the popup window from intercepting all events
            app::set_grab::<window::MenuWindow>(None);
            true
        }
        _ => false,
    });

    let mut win = window::Window::default()
        .with_size(400, 300)
        .center_screen()
        .with_label("Popup Example");
    let mut activation_frame = frame::Frame::default()
        .with_size(150, 50)
        .center_of_parent()
        .with_label("Left click for popup\nRight click for menu");
    activation_frame.set_tooltip("this is a tooltip");
    activation_frame.set_frame(enums::FrameType::BorderBox);
    activation_frame.handle(move |_, evt| match evt {
        enums::Event::Push => {
            if app::event_mouse_button() == app::MouseButton::Left {
                let (ex, ey) = (app::event_x_root(), app::event_y_root());
                popup.set_pos(ex, ey);
                popup.show();
                // allow the popup window to intercept all events
                app::set_grab(Some(popup.clone()));
            } else if app::event_mouse_button() == app::MouseButton::Right {
                let (ex, ey) = app::event_coords();
                let menu = menu::MenuItem::new(&["Item 1", "Item 2"]);
                menu.popup(ex, ey);
            }

            true
        }
        _ => false,
    });
    win.end();
    win.show();

    root.run().unwrap();
}
```

Going through my previous issues:

1. It turns out that popups have pretty extensive handlers that manually define the conditions under which the popups should close. 
2. Calling `set_override()` prevents the main window from getting pushed into the background and having a grayed out appearance.
3. Calling `app::set_grab()` allows the popup window to intercept all incoming events, which prevents the main window (or any other window for that matter) from receiving events and providing visual feedback.
4. Similarly, in the example above, calling `app::grab()` prevents the main window from receiving any events that may cause it to close. So, logically, the main window has no way of closing before the popup window. However, this is a bit fiddly, and needs special attention. If you forget to call `app::set_grab()` you would be able to send events to the main window, which makes it possible to close the main window and leave the popup hanging. Additionally, if you call `app::grab()` to allow the popup to intercept all incoming event, without a way for the popup to close and relinquish its grab on the events, you would be completely unable to interact with your desktop; since this function affects the entire windowing system.
5. `set_override()` causes the popup to get drawn as intended, with the correct drop shadow.

Here is the output of the above code:
![popup_example_2](https://user-images.githubusercontent.com/4706674/147628916-1cb0393b-7446-4e32-877f-41be6389b328.gif)


The only issue here is that the `set_override()` function is not currently exposed in the rust binding, but I'm working on a PR to change that.

**Reply by AshfordN** [reply]
See PR #1080 and [cfltk PR#211](https://github.com/MoAlyousef/cfltk/pull/211)

**Reply by MoAlyousef** [reply]
Nice. I didn't know that. Thanks for the PRs.
